Drop citations from Opinions queries if we don't have them

albertisfu opened the pull request:
As outlined in #4252, this PR introduces a feature that checks if the query contains citations. It searches for the citations in the Opinions index, and any citations that cannot be found are removed from the query. The query is then performed without the missing citations, and the user is informed accordingly.

It works for one or many missing citations in the query.

**Example with one missing citation:**
![Screenshot 2024-08-27 at 10 06 58 a m](https://github.com/user-attachments/assets/dc80ea0d-3fed-4b49-bc79-9dd6d5e3b7a3)

**Example with two missing citations:**
![Screenshot 2024-08-27 at 10 07 16 a m](https://github.com/user-attachments/assets/76dbb86f-4bb6-4f22-b5bc-8c0204b27882)

**Example with one missing citation and one that could be found.**
![Screenshot 2024-08-27 at 10 07 59 a m](https://github.com/user-attachments/assets/cc442798-5672-4dd5-9f38-b3db46694d2f)

One remaining question:
As you can see, the modified query is shown in the search bar. Or should we display the original query instead?

mlissner commented:
I think Google shows the way on this one and we should copy how it does typos:

![image](https://github.com/user-attachments/assets/ba86ca1f-616d-4eb2-bc0a-22023a3d9aab)

(We don't need the "Search Instead" part though. We can do that another day.

But you can see that Google leaves the query as it was. 

For the error, I think it'd be better if it said:

> Showing results for "Le v. Aguilar Media" without the citation "XYZ." It appears we don't yet have that citation.

And I'd change the error message to use the blue warning box too (like when we find citations).

albertisfu commented:
Thanks! I've tweaked it so that the original query is preserved in the search box.

- Updated the UI to use a blue alert box.
- Updated the alert message.

 Here some screenshots:

A single missing citation:
![Screenshot 2024-08-27 at 3 44 44 p m](https://github.com/user-attachments/assets/13dcb785-1642-44b5-813d-f0156392aca7)

Multiple missing citations:
![Screenshot 2024-08-27 at 3 43 43 p m](https://github.com/user-attachments/assets/f62b673c-a166-4cf6-8053-437797fc45ae)

Mobile:
![Screenshot 2024-08-27 at 3 44 15 p m](https://github.com/user-attachments/assets/8b2e96bc-ad82-4b5a-9f7d-d02bbde7a636)

mlissner commented on cl/search/templates/search.html:
```diff
@@ -218,6 +217,19 @@
+            <div class="row">
+                <div class="col-xs-12">
+                  <p id="missing-citations"><i class="fa fa-info-circle fa-lg"></i>
+                    Showing results for <strong>"{{ suggested_query }}"</strong> without citation{{ missing_citations|pluralize }}
+                    {% for citation in missing_citations %} <strong>"{{ citation }}"</strong>{% if not forloop.last %}, {% endif %}{% endfor %}.</br>
+                    It appears we don't yet have that citation{{ missing_citations|pluralize }}.
```
A couple tweaks:

```suggestion
                    Showing results for <strong>"{{ suggested_query }}"</strong> without citation{{ missing_citations|pluralize }}
                    {% for citation in missing_citations %} <strong>"{{ citation }}"</strong>{% if not forloop.last %}, {% endif %}{% endfor %}. It appears we don't yet have {{ missing_citations|pluralise:"that citation,those citations" }}.
```

(fixes pluralize and I don't think we need the br.)

albertisfu replied:
Thanks pluralize fixed:

![Screenshot 2024-08-27 at 4 36 41 p m](https://github.com/user-attachments/assets/658c8bd4-9303-4e3e-8665-d6cb0e4179e2)
![Screenshot 2024-08-27 at 4 35 37 p m](https://github.com/user-attachments/assets/461bfa3c-e500-46a0-9d6d-419985d9467c)

mlissner commented on cl/search/views.py:
```diff
@@ -667,6 +669,30 @@ def es_search(request: HttpRequest) -> HttpResponse:
+    :param missing_citations: A list of FullCaseCitation objects representing
+    the citations that are missing from the query.
+    :param cd: A CleanData object containing the query string.
+    :return: A twp tuple containing a list of missing citation strings and the
```
```suggestion
    :return: A two-tuple containing a list of missing citation strings and the
```

mlissner reviewed (approved):
A couple little things, but looks good otherwise. Thank you!

mlissner commented:
Lovely, thank you Alberto. Set for auto-merge! Little things like this will eventually make our results really nice.
